### PR TITLE
Linker: introduce boot and pinned regions

### PR DESCRIPTION
**Pull request by dcpleung**
This adds the bits so boot and pinned regions can be specified. These regions each contains their own text, data, rodata, bss and noinit linker sections. The boot region allows grouping of data and functions that are only used during the boot process. The associated physical page frames can be re-used if so desired with paging enabled. The pinned region hosts data and functions which must always reside in physical memory, for example, page fault handling routines for demand paging. This also provides the necessary macros to specify which section to place symbols. Currently, only a few source files are using these macros. More work needs to be done to classify at least the core kernel on where each symbol belongs.

### CODE REVIEW

**Comment by nashif** [comment]
@dcpleung  can you rebase?

**Comment by dcpleung** [comment]
Rebased

**Review by jenmwms — APPROVED** [review]
LGTM. I spot tested on ehl_crb with twister for tests/kernel/timer and /threads. Please let me know if there is a better test to exercise this.

**Review by ceolin — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by andyross — APPROVED** [review]

File: include/sys/device_mmio.h [hunk 1]
```diff
@@ -16,6 +16,7 @@
 #define ZEPHYR_INCLUDE_SYS_DEVICE_MMIO_H
 
 #include <toolchain/common.h>
+#include <linker/sections.h>
```
Comment (marc-hb) [inline include/sys/device_mmio.h:19]: In #41555 I'm moving this `#include` before the previous one to fix a BUILD_ASSERT redefinition

**Pull request merged**